### PR TITLE
Search implementation

### PR DESCRIPTION
**Pull request by ASharma2001**
Finished implementing the search functionality. The search bar now works with API function written in api_scripts.js

Also implemented Local Storage for both dietary and intolerances in the settings page. It works with the search bar API call, adding both the dietary and intolerances to the API query path. 

Added ADRs for both the database and API. 

Added documentation file for Local Storage

### CODE REVIEW

**Comment by Dsuzuki126** [comment]
I think it looks like it works,
There are only two things I am worried about
1) the intolerances and dietary restrictions not saving or updating properly as you change pages
2) the intolerances/dietary restrictions not applying to the results shown

File: source/scripts/api_script.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+// Fetch the recipes for the search page
+// queries - string that contains all query parameters for API call
+// callbackFn assigns the value from the fetch to the global variable
+async function fetchRecipes(queries, callbackFn) {
+    const complexSearch = "/complexSearch";
+    const url = `${MAIN_API_URL}${complexSearch}?${queries}${API_KEY}`;
+    console.log(url);
+    return new Promise((resolve, reject) => {
+        fetch(url)
+        .then(response => response.json())
```
Comment (kjl001) [inline source/scripts/api_script.js:31]: Codacy has a fix for the issue: [Expected parentheses around arrow function argument.](https://app.codacy.com/gh/cse110-fa21-group23/cse110-fa21-group23/pullRequest?prid=8408391)
```suggestion
        .then((response) => response.json())
```

File: source/scripts/script.js [hunk 2]
```diff
@@ -19,54 +16,63 @@ async function init() {
...
-
+// The search function, calls API function to fetch all recipes
+// Generates recipe cards by passing in values into RecipeData
 function search() {
-    // let searchQuery = document.getElementById('search-query').value;
-    // console.log(searchQuery);
-    // console.log(localStorage.getItem("dietaryRestrictions"));
-    //hideCategoryCards();
+    // get the search query
+    const searchQuery = document.getElementById('search-query').value;
```
Comment (kjl001) [inline source/scripts/script.js:36]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-fa21-group23/cse110-fa21-group23/pullRequest?prid=8408391)
```suggestion
    const searchQuery = document.getElementById("search-query").value;
```

File: source/scripts/api_script.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+ * @returns {{diets1,id1,image1,title1},{diets2,id2,image2,title2},...}
+ */
+function getRecipeCardInfo(data) {
+    const recipes = [];
+    for (let i = 0; i < data.length; i++) {
+        const result = {};
+        result['id'] = data[i]['id'];
+        result['title'] = data[i]['title'];
+        result['image'] = data[i]['image'];
+        result['diets'] = data[i]['diets'];
```
Comment (kjl001) [inline source/scripts/api_script.js:56]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-fa21-group23/cse110-fa21-group23/pullRequest?prid=8408391)
```suggestion
        result["diets"] = data[i]['diets'];
```

File: source/scripts/script.js [hunk 4]
```diff
@@ -19,54 +16,63 @@ async function init() {
         }
     });
 
+    // Add click event listener for search button
+    const searchButton = document.getElementById('search-button');
+    searchButton.addEventListener('click', async () => {
```
Comment (kjl001) [inline source/scripts/script.js:21]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-fa21-group23/cse110-fa21-group23/pullRequest?prid=8408391)
```suggestion
    searchButton.addEventListener("click", async () => {
```

File: source/scripts/script.js [hunk 5]
```diff
@@ -37,51 +32,67 @@ async function init() {
...
     // bindShowMore();
-
 }
 
-
+// The search function, calls API function to fetch all recipes
+// Generates recipe cards by passing in values into RecipeData
 function search() {
+    // get the search query
+    const searchQuery = document.getElementById('search-query').value;
```
Comment (kjl001) [inline source/scripts/script.js:52]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-fa21-group23/cse110-fa21-group23/pullRequest?prid=8408391)
```suggestion
    const searchQuery = document.getElementById("search-query").value;
```

File: source/scripts/api_script.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    return new Promise((resolve, reject) => {
+        fetch(url)
+        .then((response) => response.json())
+        .then(data => {
+                const recipes = getRecipeCardInfo(data.results);
+                callbackFn(data.results);
+                resolve(true);
+            })
+        .catch(err => {
+                console.log(`Error loading the recipe`);
```
Comment (kjl001) [inline source/scripts/api_script.js:39]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-fa21-group23/cse110-fa21-group23/pullRequest?prid=8408391)
```suggestion
                console.log("Error loading the recipe");
```

File: source/scripts/api_script.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ * Extract needed info of all recipes fetched as an array 
+ * @returns {{diets1,id1,image1,title1},{diets2,id2,image2,title2},...}
+ */
+function getRecipeCardInfo(data) {
+    const recipes = [];
+    for (let i = 0; i < data.length; i++) {
+        const result = {};
+        result["id"] = data[i]['id'];
+        result["title"] = data[i]['title'];
```
Comment (kjl001) [inline source/scripts/api_script.js:55]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-fa21-group23/cse110-fa21-group23/pullRequest?prid=8408391)
```suggestion
        result["title"] = data[i]["title"];
```

File: source/scripts/script.js [hunk 8]
```diff
@@ -14,20 +9,20 @@ const images = ["./img/foodPics/indian.jpeg", "./img/foodPics/vegan.jpeg", "./im
...
-    }
-});
-
 window.addEventListener('DOMContentLoaded', init);
 
 async function init() {
     showHome();
     createCategoryCards();
     showRecipePage();
+    const clearBtn = document.getElementById('clear-btn');
```
Comment (kjl001) [inline source/scripts/script.js:18]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-fa21-group23/cse110-fa21-group23/pullRequest?prid=8408391)
```suggestion
    const clearBtn = document.getElementById("clear-btn");
```

File: source/scripts/script.js [hunk 9]
```diff
@@ -101,17 +112,28 @@ function showSettings() {
     const settings = document.getElementById("settings-container");
     settings.style.visibility = "visible";
     //settings.style.transform = "translate(100%)";
+
     // Get the list of restrictions from local storage
     const getDietaryRestrictions = JSON.parse(localStorage.getItem("dietaryRestrictions"));
+    const getIntolerancesRestrictions = JSON.parse(localStorage.getItem("intolerancesRestrictions"));
     const dietaryContainerElements = document.getElementById('dietary-container').elements;
+    const intolerancesContainerElements = document.getElementById('intolerances-container').elements;
```
Comment (kjl001) [inline source/scripts/script.js:120]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-fa21-group23/cse110-fa21-group23/pullRequest?prid=8408391)
```suggestion
    const intolerancesContainerElements = document.getElementById("intolerances-container").elements;
```

File: source/scripts/script.js [hunk 10]
```diff
@@ -1,10 +1,5 @@
-const API_KEY = '8aaa6b0816db4a99b92e7852d125a9aa';
-// API_KEY4 (Nhi): 8aaa6b0816db4a99b92e7852d125a9aa
-// API_KEY3 (Nhi): c8f83bb3a9af4355b12de10250b24c88
-// API_KEY2 (Nhi): fafd5e810c304ed3b4f9984672cb21ee
-// API_KEY1: 4d936c811cda46879d4749def6bb36a1
-const url = `https://api.spoonacular.com/recipes/complexSearch?apiKey=${API_KEY}&addRecipeInformation=true&instructionsRequired=true`;
-const recipes = [];
+// Import API from file
+import { fetchRecipes } from '/source/scripts/api_script.js';
```
Comment (kjl001) [inline source/scripts/script.js:2]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/cse110-fa21-group23/cse110-fa21-group23/pullRequest?prid=8408391)
```suggestion
import { fetchRecipes } from "/source/scripts/api_script.js";
```

**Pull request merged**